### PR TITLE
parser/encoding/rpc: refactor encoding model

### PR DESCRIPTION
**Pull request by ekerfelt**
Refactor the encoding model in preparation for api explorer changes

### CODE REVIEW

File: cli/internal/codegen/testdata/expected_golang.go [hunk 1]
```diff
@@ -371,10 +371,10 @@ func (c *svcClient) RequestWithAllInputTypes(ctx context.Context, params SvcAllI
 	// Copy the unmarshalled response body into our response struct
 	respDecoder := &serde{}
 
-	resp.A = respDecoder.ToTime("A", respHeaders.Get("x-alice"), false)
 	resp.B = respBody.B
 	resp.C = respBody.C
 	resp.Dave = respBody.Dave
+	resp.A = respDecoder.ToTime("A", respHeaders.Get("x-alice"), false)
```
Comment (eandre) [inline cli/internal/codegen/testdata/expected_golang.go:377]: Why the ordering change? Is it due to the refactoring or is the output indeterministic?

**Review by eandre — COMMENTED** [review]
Looks great to me but leaving for @DomBlack to approve. Just one minor question

File: cli/internal/codegen/testdata/expected_golang.go (hunk 1 above)
Comment (ekerfelt) [reply cli/internal/codegen/testdata/expected_golang.go:377]: It's due to the refactoring. I can change it back to make it consistent

File: parser/encoding/rpc.go [hunk 2]
```diff
@@ -13,6 +13,14 @@ import (
...
 )
 
+type Lang int
+
+const (
+	_ Lang = iota
+	GO
+	TypeScript
+)
+
```
Comment (DomBlack) [inline parser/encoding/rpc.go:23]: I'm not sure I agree with moving language-specific logic in here; I'd rather we kept `SrcField` as it appears in the Go code, and then let any language generator make any changes it deems are needed.

For instance, if we had a Swift client, that will have a different naming convention to TypeScript and Go - does it mean that the `encoding/rpc.go` file now also needs to understand that as well?

Instead if we keep it `SrcField` as the Go struct field name; then the various client generators can update as needed

File: cli/internal/codegen/typescript.go [hunk 3]
```diff
@@ -356,16 +353,10 @@ func (ts *typescript) rpcCallSite(ns string, w *indentWriter, rpc *meta.RPC, rpc
 
 	w.WriteStringf("// Now make the actual call to the API\nconst resp = await %s\n", callAPI)
 
-	headerFields, queryFields, _, err := toFieldLists(rpcEncoding.ResponseEncoding.Fields)
-	if err != nil {
-		return err
-	}
-	if len(queryFields) > 0 {
```
[the remark this answers is not in the document]
Comment (ekerfelt) [reply cli/internal/codegen/typescript.go:363]: Fair, we should move it into the DescribeRespose method

File: parser/encoding/rpc.go (hunk 2 above)
Comment (ekerfelt) [reply parser/encoding/rpc.go]: Agreed with Dom on a solution

**Review by DomBlack — DISMISSED** [review]

**Review by DomBlack — APPROVED** [review]

**Pull request merged**